PPS global alignment - reformat

MatiXOfficial opened the pull request:
#### PR description:

This is the last planned PR for PPS global alignment. It does not change the logic of the code nor anything in the PCL workflow configuration and the matrix test. The main purpose is reformatting the code, updating the README files and adding one more parameter to `PPSAlignmentHarvester` - reasonability range - which ensures that the harvester will not produce NaN or extremely big alignment corrections in the output.

Apart from the CMSSW changes, a [new TWiki page](https://twiki.cern.ch/twiki/bin/viewauth/CMS/PPSAlign) was created, and some part of the documentation that previously had been in the README files was moved there.

For further information about including global alignment in CMSSW check out two previous PRs: #35631 and #35874.

<!-- Please replace this text with a description of the feature proposed or problem addressed, specifying:
  - what changes are expected in the output if any, 
  - what other PRs or externals it depends upon if any,
  - link to any additional material useful to provide a documentation for this PR (slides, JIRA tickets, related pull requestes, hypernews, TWiki or Indico pages)  -->

#### PR validation:

- The `1042` relval runs correctly. 
- The test from *CalibPPS/AlignmentGloba/test* reproduced the results. 
- Another tests based on Run 2 data also reproduced results.

MatiXOfficial committed:
Small reformat

- Replaced some raw pointers with the smart ones in PPSAlignmentConfigurationESSource.
- Simplified pair iteration in PPSAlignmentHarvester, xAlignment.

MatiXOfficial committed:
Moved fitProfile to utils.h.

MatiXOfficial committed:
Added reasonability ranges to the harvester.

If the corrections created by the harvester are not compatible with these ranges, they are replaced with 0.

MatiXOfficial committed:
Replaced folder variables with the ones with more reasonable names.

MatiXOfficial committed:
AlignmentGlobal package reformat

- Updated the names of some variables.
- Improved formatting in a few places.
- Modified some comments and added new ones.
- Removed unnecessary imports.
- Added const where applicable.
- Replaced some function parameters with class/struct members.

MatiXOfficial committed:
Moved a small part of the code to a new SlicePlots method.

MatiXOfficial committed:
Updated the READMEs in AlignmentGlobal.

MatiXOfficial committed:
Changed the tag in the AlignmentGlobal test config.

MatiXOfficial committed:
Added comments to the PPSAlignmentConfig classes to mark them obsolete.

cmsbuild commented:
+code-checks

Logs: https://cmssdt.cern.ch/SDT/code-checks/cms-sw-PR-36257/26916

- This PR adds an extra 52KB to repository

cmsbuild commented:
A new Pull Request was created by  @MatiXOfficial (Mateusz Kocot) for master.

It involves the following packages:

- CalibPPS/AlignmentGlobal (**alca**)
- CalibPPS/ESProducers (**alca**)
- CondFormats/PPSObjects (**alca**)


@cmsbuild, @malbouis, @tvami, @yuanchao, @francescobrivio can you please review it and eventually sign? Thanks.
@tocheng, @fabferro, @jan-kaspar, @mmusich, @seemasharmafnal this is something you requested to watch as well.
@perrotta, @dpiparo, @qliphy you are the release manager for this.

cms-bot commands are listed <a href="http://cms-sw.github.io/cms-bot-cmssw-cmds.html">here</a>

malbouis commented:
test parameters:

- workflows = 1042

malbouis commented:
@cmsbuild , please test

cmsbuild commented:
+1

**Summary**: https://cmssdt.cern.ch/SDT/jenkins-artifacts/pull-request-integration/PR-ce6a23/20788/summary.html
**COMMIT**: 88d93fd550059475d64497de58ba299ccd339bb1
**CMSSW**: CMSSW_12_2_X_2021-11-26-1100/slc7_amd64_gcc900
**User test area**: For local testing, you can use `/cvmfs/cms-ci.cern.ch/week0/cms-sw/cmssw/36257/20788/install.sh` to create a dev area with all the needed externals and cmssw changes.


## Comparison Summary



@slava77 comparisons for the following workflows were not done due to missing matrix map:
   * /data/cmsbld/jenkins/workspace/compare-root-files-short-matrix/data/PR-ce6a23/1042.0_RunExpressPhy2017F+RunExpressPhy2017F+TIER0EXPPPSCALALIG+ALCASPLITPPSALIG+ALCAHARVDPPSALIG

**Summary**:
 - No significant changes to the logs found
 - Reco comparison results: 4 differences found in the comparisons
 - DQMHistoTests: Total files compared: 42
 - DQMHistoTests: Total histograms compared: 3247873
 - DQMHistoTests: Total failures: 5
 - DQMHistoTests: Total nulls: 1
 - DQMHistoTests: Total successes: 3247845
 - DQMHistoTests: Total skipped: 22
 - DQMHistoTests: Total Missing objects: 0
 - DQMHistoSizes: Histogram memory added: -0.004 KiB( 41 files compared)
 - DQMHistoSizes: changed ( 312.0 ): -0.004 KiB       MessageLogger/Warnings
 - Checked 177 log files, 37 edm output root files, 42 DQM output files
 - [TriggerResults](https://cmssdt.cern.ch/SDT/jenkins-artifacts/baseLineComparisons/CMSSW_12_2_X_2021-11-26-1100+ce6a23/47079/triggerResults): no differences found

tvami commented on CalibPPS/ESProducers/plugins/PPSAlignmentConfigESSource.cc:
```diff
@@ -1,8 +1,6 @@
- *  CalibPPS/ESProducers/plugins/PPSAlignmentConfigESSource.cc
- * 
- *  Description: Constructs PPSAlignmentConfig instance
+ *  This module is obsolete! Use PPSAlignmentConfigurationESSource instead.
```
Should there be a real exception thrown if somebody wants to use this? I'm not sure a single comment in the code is sufficient to stop people to use it haha

MatiXOfficial replied:
I just wanted to make sure no one is confused when they see two very similar classes in one directory, but throwing an exception is a good idea too. I'll add it.

mmusich replied:
if the module is obsolete, what's the reason to keep it in the code-base? 
Are there still clients of it?

MatiXOfficial replied:
`PPSAlignmentConfiguration` has fully replaced `PPSAlignmentConfig`, so nobody will use its ESSource in the future. To be honest, I wanted to delete all the `PPSAlignmentConfig` auxiliary classes in one of the previous PRs (https://github.com/cms-sw/cmssw/pull/35174#discussion_r704723640), but eventually we decided to keep them. Do you suggest to delete them now?

mmusich replied:
the less code is in the stack, the less maintenance it will require. If the functionality is fully replaced by other modules, deleting the obsolete files sounds like a good option (and will also automatically enforce that nobody uses them by mistake).

MatiXOfficial replied:
Ok then, I'll delete them. Am I right though that I can't delete the `PPSAlignmentConfig` class? It has a few payloads in the DB.

mmusich replied:
>  Am I right though that I can't delete the PPSAlignmentConfig class? It has a few payloads in the DB.

provided they are not included in any `autoCond` GlobalTag you technically *could* delete the C++ class as well, though if I am not mistaken AlCa/DB discourages that practice @cms-sw/alca-l2 can comment further.

MatiXOfficial replied:
Ok, thanks! I will leave the class for now.

tvami commented on CondFormats/PPSObjects/src/PPSAlignmentConfig.cc:
```diff
@@ -1,8 +1,6 @@
- *  CondFormats/PPSObjects/interface/PPSAlignmentConfig.cc
- *
- *  Description : Class with alignment parameters
+ *  This class is obsolete! Use PPSAlignmentConfiguration instead.
```
Same comment as in https://github.com/cms-sw/cmssw/pull/36257/files#r757717599

tvami commented on CalibPPS/AlignmentGlobal/plugins/PPSAlignmentHarvester.cc:
```diff
@@ -232,18 +236,19 @@ void PPSAlignmentHarvester::dqmEndRun(DQMStore::IBooker& iBooker,
       doYAli = true;
     } else
-      edm::LogError("PPS") << "[harvester] " << sequence_[i] << " is a wrong method name.";
+      edm::LogError("PPS") << "[harvester] " << aliMethod << " is a wrong method name.";
```
```suggestion
      edm::LogError("PPSAlignmentHarvester") << aliMethod << " is a wrong method name.";
```
In the parentheses of the LogError you need to put the class name

MatiXOfficial replied:
Ok, I'll change the log categories to the proper classes. I remember that once we've decided to use "PPS" everywhere, but now I don't remember why ;)

tvami commented on CalibPPS/AlignmentGlobal/plugins/PPSAlignmentHarvester.cc:
```diff
@@ -770,19 +756,19 @@ void PPSAlignmentHarvester::xAlignmentRelative(DQMStore::IBooker& iBooker,
 
   // write results
-  edm::LogInfo("PPS") << seqPos + 1 << ": x_alignment_relative:\n"
+  edm::LogInfo("PPS") << seqPos << ": x_alignment_relative:\n"
```
```suggestion
  edm::LogInfo("PPSAlignmentHarvester") << seqPos << ": x_alignment_relative:\n"

```

tvami commented on CalibPPS/AlignmentGlobal/plugins/PPSAlignmentHarvester.cc:
```diff
@@ -982,19 +970,21 @@ void PPSAlignmentHarvester::yAlignment(DQMStore::IBooker& iBooker,
   // write results
-  edm::LogInfo("PPS") << seqPos + 1 << ": y_alignment:\n"
-                      << yAliResults_ << seqPos + 1 << ": y_alignment_sl_fix:\n"
+  edm::LogInfo("PPS") << seqPos << ": y_alignment:\n"
```
```suggestion
  edm::LogInfo("PPSAlignmentHarvester") << seqPos << ": y_alignment:\n"

```

tvami commented:
Besides the cosmetics above, the DQM output https://cmssdt.cern.ch/SDT/jenkins-artifacts/pull-request-integration/PR-ce6a23/20788/runTheMatrix-results/1042.0_RunExpressPhy2017F+RunExpressPhy2017F+TIER0EXPPPSCALALIG+ALCASPLITPPSALIG+ALCAHARVDPPSALIG/DQM_V0001_R000305064__Express__PCLTest__ALCAPROMPT.root looks good to me

MatiXOfficial committed:
Deleted obsolete files and lines of code related with PPSAlignmentConfig.

MatiXOfficial committed:
Updated logging categories in the AlignmentGlobal plugins.

MatiXOfficial commented:
@tvami @mmusich I've added your suggestions. When it comes to `PPSAlignmentConfig`, I deleted its auxiliary classes and lines of code, but I left the class itself with the `REGISTER_PLUGIN` and `TYPELOOKUP_DATA_REG` macros.

cmsbuild commented:
+code-checks

Logs: https://cmssdt.cern.ch/SDT/code-checks/cms-sw-PR-36257/26939

- This PR adds an extra 224KB to repository

cmsbuild commented:
Pull request #36257 was updated. @malbouis, @yuanchao, @cmsbuild, @ggovi, @francescobrivio, @tvami can you please check and sign again.

tvami commented:
@cmsbuild , please test

cmsbuild commented:
+1

**Summary**: https://cmssdt.cern.ch/SDT/jenkins-artifacts/pull-request-integration/PR-ce6a23/20796/summary.html
**COMMIT**: 5899d3c4c0497d7ed422bc52a0ea00b825c648e4
**CMSSW**: CMSSW_12_2_X_2021-11-27-1100/slc7_amd64_gcc900
**User test area**: For local testing, you can use `/cvmfs/cms-ci.cern.ch/week0/cms-sw/cmssw/36257/20796/install.sh` to create a dev area with all the needed externals and cmssw changes.


## Comparison Summary



@slava77 comparisons for the following workflows were not done due to missing matrix map:
   * /data/cmsbld/jenkins/workspace/compare-root-files-short-matrix/data/PR-ce6a23/1042.0_RunExpressPhy2017F+RunExpressPhy2017F+TIER0EXPPPSCALALIG+ALCASPLITPPSALIG+ALCAHARVDPPSALIG

**Summary**:
 - No significant changes to the logs found
 - Reco comparison results: 0 differences found in the comparisons
 - DQMHistoTests: Total files compared: 42
 - DQMHistoTests: Total histograms compared: 3247873
 - DQMHistoTests: Total failures: 0
 - DQMHistoTests: Total nulls: 0
 - DQMHistoTests: Total successes: 3247851
 - DQMHistoTests: Total skipped: 22
 - DQMHistoTests: Total Missing objects: 0
 - DQMHistoSizes: Histogram memory added: 0.0 KiB( 41 files compared)
 - Checked 177 log files, 37 edm output root files, 42 DQM output files
 - [TriggerResults](https://cmssdt.cern.ch/SDT/jenkins-artifacts/baseLineComparisons/CMSSW_12_2_X_2021-11-27-1100+ce6a23/47086/triggerResults): no differences found

tvami commented:
+alca
 * PR description in line w code changes
 * tests pass, the wf 1042 shows non-empty DQM plots (well most of the time)

ggovi commented on CondCore/Utilities/plugins/Module_2XML.cc:
```diff
@@ -31,7 +31,6 @@ PAYLOAD_2XML_MODULE(pluginUtilities_payload2xml) {
   PAYLOAD_2XML_CLASS(CTPPSPixelAnalysisMask);
   PAYLOAD_2XML_CLASS(CTPPSPixelDAQMapping);
   PAYLOAD_2XML_CLASS(CTPPSPixelGainCalibrations);
-  PAYLOAD_2XML_CLASS(PPSAlignmentConfig)
```
@MatiXOfficial 
As Far as I understand, you have kept the old condformat (PPSAlignmentConfig). In this case, better also keep the functionalities to dump the corresponding payloads in the database. This requires the PAYLOAD_2XML_CLASS macro entry.

MatiXOfficial replied:
Ok, I restored the macro and the test.

ggovi commented on CondCore/Utilities/src/CondFormats.h:
```diff
@@ -36,7 +36,6 @@
 #include "CondFormats/PPSObjects/interface/PPSTimingCalibration.h"
 #include "CondFormats/PPSObjects/interface/PPSAlignmentConfiguration.h"
 #include "CondFormats/PPSObjects/interface/LHCOpticalFunctionsSetCollection.h"
-#include "CondFormats/PPSObjects/interface/PPSAlignmentConfig.h"
```
@MatiXOfficial See comment above - required by the conddb dump command

ggovi commented on CondFormats/PPSObjects/test/BuildFile.xml:
```diff
@@ -10,10 +10,6 @@
   <use name="CondFormats/PPSObjects"/>
 </bin>
 
-<bin file="testSerializationPPSAlignmentConfig.cc">
```
@MatiXOfficial You can keep this one...

tvami commented:
@cmsbuild , please test

cmsbuild commented:
+code-checks

Logs: https://cmssdt.cern.ch/SDT/code-checks/cms-sw-PR-36257/27147

- This PR adds an extra 16KB to repository

- Found files with invalid states:
  - CondFormats/PPSObjects/test/testSerializationPPSAlignmentConfig.cc:
    - Deleted: 6e3020075a7ce5151e81c63d64768e8fe7e36fb5
    - Added: 73bb67838a91cc83901591117c26be2bd693d9bc

cmsbuild commented:
Pull request #36257 was updated. @ggovi, @malbouis, @tvami, @yuanchao, @francescobrivio can you please check and sign again.

tvami commented:
>     * Deleted: [6e30200](https://github.com/cms-sw/cmssw/commit/6e3020075a7ce5151e81c63d64768e8fe7e36fb5)
>     * Added: [73bb678](https://github.com/cms-sw/cmssw/commit/73bb67838a91cc83901591117c26be2bd693d9bc)

@MatiXOfficial please squash the commit that create and delete these files so we can rid of this virtual contribution

MatiXOfficial commented:
> > ```
> > * Deleted: [6e30200](https://github.com/cms-sw/cmssw/commit/6e3020075a7ce5151e81c63d64768e8fe7e36fb5)
> > * Added: [73bb678](https://github.com/cms-sw/cmssw/commit/73bb67838a91cc83901591117c26be2bd693d9bc)
> > ```
> 
> @MatiXOfficial please squash the commit that create and delete these files so we can rid of this virtual contribution

@tvami Done

cmsbuild commented:
+code-checks

Logs: https://cmssdt.cern.ch/SDT/code-checks/cms-sw-PR-36257/27149

- This PR adds an extra 16KB to repository

cmsbuild commented:
Pull request #36257 was updated. @malbouis, @yuanchao, @cmsbuild, @ggovi, @francescobrivio, @tvami can you please check and sign again.

tvami commented:
@cmsbuild , please test

cmsbuild commented:
+1

**Summary**: https://cmssdt.cern.ch/SDT/jenkins-artifacts/pull-request-integration/PR-ce6a23/21024/summary.html
**COMMIT**: 21fb68d9b01202eaca5b70361b43ddda521640f8
**CMSSW**: CMSSW_12_2_X_2021-12-06-1100/slc7_amd64_gcc900
**User test area**: For local testing, you can use `/cvmfs/cms-ci.cern.ch/week0/cms-sw/cmssw/36257/21024/install.sh` to create a dev area with all the needed externals and cmssw changes.


## Comparison Summary



@slava77 comparisons for the following workflows were not done due to missing matrix map:
   * /data/cmsbld/jenkins/workspace/compare-root-files-short-matrix/data/PR-ce6a23/1042.0_RunExpressPhy2017F+RunExpressPhy2017F+TIER0EXPPPSCALALIG+ALCASPLITPPSALIG+ALCAHARVDPPSALIG

**Summary**:
 - No significant changes to the logs found
 - Reco comparison results: 4 differences found in the comparisons
 - DQMHistoTests: Total files compared: 42
 - DQMHistoTests: Total histograms compared: 3250608
 - DQMHistoTests: Total failures: 6
 - DQMHistoTests: Total nulls: 0
 - DQMHistoTests: Total successes: 3250580
 - DQMHistoTests: Total skipped: 22
 - DQMHistoTests: Total Missing objects: 0
 - DQMHistoSizes: Histogram memory added: 0.0 KiB( 41 files compared)
 - Checked 177 log files, 37 edm output root files, 42 DQM output files
 - [TriggerResults](https://cmssdt.cern.ch/SDT/jenkins-artifacts/baseLineComparisons/CMSSW_12_2_X_2021-12-06-1100+ce6a23/47271/triggerResults): no differences found

tvami commented:
+1
 * code change in line w PR description and code review
 * Jenkins test pass

cmsbuild commented:
This pull request is fully signed and it will be integrated in one of the next master IBs (tests are also fine). This pull request will now be reviewed by the release team before it's merged. @perrotta, @dpiparo, @qliphy (and backports should be raised in the release meeting by the corresponding L2)

tvami commented on CondCore/Utilities/src/CondDBFetch.cc:
```diff
@@ -56,7 +56,6 @@ namespace cond {
       FETCH_PAYLOAD_CASE(CTPPSPixelAnalysisMask)
       FETCH_PAYLOAD_CASE(CTPPSPixelGainCalibrations)
       FETCH_PAYLOAD_CASE(CTPPSRPAlignmentCorrectionsData)
```
@ggovi is this deletion fine?

ggovi replied:
@tvami It is only relevant for conddb_import, which is an obsolete command anyhow... Therefore yes, it is fine.

tvami commented:
+db

cmsbuild commented:
This pull request is fully signed and it will be integrated in one of the next master IBs (tests are also fine). This pull request will now be reviewed by the release team before it's merged. @perrotta, @dpiparo, @qliphy (and backports should be raised in the release meeting by the corresponding L2)

qliphy commented:
+1
